Zend/zend_execute.c

```diff
@@ -5253,14 +5257,158 @@ static zend_always_inline zend_execute_data *_zend_vm_stack_push_call_frame(uint
 # include "zend_vm_trace_map.h"
 #endif
 
+#ifdef ZEND_VERIFY_TYPE_INFERENCE
+static bool zend_verify_type_inference(uint32_t type_mask, zval *value)
+{
+	if (type_mask == 0) {
+		return true;
+	}
+
+	// IS_PTR is used internally, skip for now
+	if (Z_TYPE_P(value) == IS_PTR) {
+		return true;
+	}
+
+	if (Z_TYPE_P(value) == IS_INDIRECT) {
+		if (!(type_mask & MAY_BE_INDIRECT)) {
+			return false;
+		}
+		value = Z_INDIRECT_P(value);
+	}
+
+	if (Z_REFCOUNTED_P(value)) {
+		if (Z_REFCOUNT_P(value) == 1 && !(type_mask & MAY_BE_RC1)) {
+			return false;
+		}
+		if (Z_REFCOUNT_P(value) > 1 && !(type_mask & MAY_BE_RCN)) {
+			return false;
+		}
+	}
+
+	if (Z_TYPE_P(value) == IS_REFERENCE) {
+		if (!(type_mask & MAY_BE_REF)) {
+			return false;
+		}
+		value = Z_REFVAL_P(value);
+	}
+
+	uint32_t type = 1u << Z_TYPE_P(value);
+	if (!(type_mask & type)) {
+		return false;
+	}
+
+	if (Z_TYPE_P(value) == IS_ARRAY) {
+		HashTable *ht = Z_ARRVAL_P(value);
+		uint32_t num_checked = 0;
+		zend_string *str;
+		zval *val;
+		ZEND_HASH_FOREACH_STR_KEY_VAL(ht, str, val) {
```

Comment:
You should also check HASH vs PACKED.
You may skip elements checking if all ``MAY_BE_ARRAY_KEY_ANY|MAY_BE_ARRAY_OF_ANY|MAY_BE_ARRAY_OF_REF`` bits are set.